src/backing/client/grpc_client/grpc_client_test.cc

```diff
@@ -0,0 +1,198 @@
+using ::kmsengine::testing_util::MockKeyManagementServiceStub;
+using ::testing::Combine;
+using ::testing::DoAll;
+using ::testing::Not;
+using ::testing::Return;
+using ::testing::SaveArg;
+using ::testing::SetArgPointee;
+using ::testing::StrEq;
+using ::testing::ValuesIn;
+
+constexpr DigestCase kDigestCases[] = {
+  DigestCase::kSha256,
+  DigestCase::kSha384,
+  DigestCase::kSha512,
+};
+
+const std::string kSampleDigests[] = {
+  // Example SHA-256 digest of "hello world" for testing.
+  "b94d27b9934d3e08a52e52d7da7dabfac484efe37a5380ee9088f7ace2efcde9",
+  "an arbitrary digest",
+  "",
+};
+
+const std::string kSampleKeyNames[] = {
+  "/projects/PROJECT_ID/locations/LOCATION/keyRings/KEY_RING/cryptoKeys/KEY"
+       "/cryptoKeyVersions/VERSION",
+  "another arbitrary key name",
+  "",
+};
+
+class AsymmetricSignTest : public
+    testing::TestWithParam<std::tuple<std::string, DigestCase, std::string>> {
+  // Purposely empty; no fixtures to instantiate.
+};
+
+INSTANTIATE_TEST_SUITE_P(AsymmetricSignParameters, AsymmetricSignTest,
+                         Combine(ValuesIn(kSampleKeyNames),
+                                 ValuesIn(kDigestCases),
+                                 ValuesIn(kSampleDigests)));
+
+TEST_P(AsymmetricSignTest, AsymmetricSignReturnsSignatureInResponse) {
+  const std::string expected_signature = "my signature";
+  const std::string key = std::get<0>(GetParam());
+  const DigestCase digest_case = std::get<1>(GetParam());
+  const std::string digest_bytes = std::get<2>(GetParam());
+
+  google::cloud::kms::v1::AsymmetricSignResponse mock_response;
+  mock_response.set_signature(expected_signature);
+
+  auto stub = absl::make_unique<MockKeyManagementServiceStub>();
+  EXPECT_CALL(*stub, AsymmetricSign)
+      .WillOnce(DoAll(SetArgPointee</*response_param_index=*/2>(mock_response),
+                      Return(Status())));
+
+  auto context_factory = absl::make_unique<MockClientContextFactory>();
+  EXPECT_CALL(*context_factory, MakeContext);
+
+  GrpcClient client(std::move(stub), std::move(context_factory));
+  StatusOr<std::string> actual = client.AsymmetricSign(key, digest_case,
+                                                       digest_bytes);
+  EXPECT_THAT(actual, IsOk());
+  EXPECT_THAT(actual.value(), StrEq(expected_signature));
+}
+
+TEST_P(AsymmetricSignTest, AsymmetricSignSetsCorrectRequestFields) {
+  const std::string expected_key = std::get<0>(GetParam());
+  const DigestCase expected_digest_case = std::get<1>(GetParam());
+  const std::string expected_digest_bytes = std::get<2>(GetParam());
+
+  google::cloud::kms::v1::AsymmetricSignRequest actual_request;
+  auto stub = absl::make_unique<MockKeyManagementServiceStub>();
+  EXPECT_CALL(*stub, AsymmetricSign)
+      .WillOnce(DoAll(SaveArg</*request_param_index=*/1>(&actual_request),
+                      Return(Status())));
+
+  auto context_factory = absl::make_unique<MockClientContextFactory>();
```

Comment:
The use of a context seems like an implementation detail that's not really the focus of these tests.. If that's the case, it might make sense to remove the mock and the expectation on `MakeContext`.

Reply:
I think checking the `MakeContext` call here is useful since it's checking that the `GrpcClient` is actually using the `ClientContextFactory` that was passed in. Previously, the implementation looked something like:

```
grpc::ClientContext context;
client->Sign(&context, ..., ...);
```

and so if someone was just creating a `grpc::ClientContext` in the implementation without using the `ClientContextFactory`, an issue occurs. This at least checks that the `ClientContextFactory` is being used during the call.

Reply:
Ack